internal/sync/cache.go

```diff
@@ -0,0 +1,471 @@
+	numaLogger := logger.New().WithName("live state cache")
+	return &liveStateCache{
+		cluster: cluster,
+		logger:  numaLogger,
+	}
+}
+
+func NewLiveStateCache(
+	clusterCacheConfig *rest.Config,
+) LiveStateCache {
+	return &liveStateCache{
+		clusterCacheConfig: clusterCacheConfig,
+	}
+}
+
+func (c *liveStateCache) invalidate(cacheSettings cacheSettings) {
+	c.logger.Info("invalidating live state cache")
+	c.lock.Lock()
+	c.cacheSettings = cacheSettings
+	cluster := c.cluster
+	c.lock.Unlock()
+
+	cluster.Invalidate(clustercache.SetSettings(cacheSettings.clusterSettings))
+
+	c.logger.Info("live state cache invalidated")
+}
+
+func (c *liveStateCache) getCluster() clustercache.ClusterCache {
+	// Load the cache setting to see if it changed
+	nextCacheSettings, err := c.loadCacheSettings()
+	if err != nil {
+		c.logger.Error(err, "failed to read cache settings")
+	}
+
+	c.lock.Lock()
+	needInvalidate := false
+	if !reflect.DeepEqual(c.cacheSettings, *nextCacheSettings) {
+		c.cacheSettings = *nextCacheSettings
+		needInvalidate = true
+	}
+	c.lock.Unlock()
+	if needInvalidate {
+		c.invalidate(*nextCacheSettings)
+	}
+
+	c.lock.RLock()
+	cluster := c.cluster
+	cacheSettings := c.cacheSettings
+	c.lock.RUnlock()
+
+	if cluster != nil {
+		return cluster
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.cluster != nil {
+		return c.cluster
+	}
+
+	clusterCacheConfig := c.clusterCacheConfig
+
+	// TODO: check if we need set managed namespace
+	clusterCacheOpts := []clustercache.UpdateSettingsFunc{
+		clustercache.SetListSemaphore(semaphore.NewWeighted(clusterCacheListSemaphoreSize)),
+		clustercache.SetListPageSize(clusterCacheListPageSize),
+		clustercache.SetListPageBufferSize(clusterCacheListPageBufferSize),
+		clustercache.SetWatchResyncTimeout(clusterCacheWatchResyncDuration),
+		clustercache.SetClusterSyncRetryTimeout(clusterSyncRetryTimeoutDuration),
+		clustercache.SetResyncTimeout(clusterCacheResyncDuration),
+		clustercache.SetSettings(cacheSettings.clusterSettings),
+		clustercache.SetPopulateResourceInfoHandler(c.PopulateResourceInfo),
+		clustercache.SetRetryOptions(clusterCacheAttemptLimit, clusterCacheRetryUseBackoff, isRetryableError),
+		clustercache.SetLogr(*logger.New().LogrLogger),
+	}
+
+	clusterCache := clustercache.NewClusterCache(clusterCacheConfig, clusterCacheOpts...)
+
+	// Register event handler that is executed whenever the resource gets updated in the cache
+	_ = clusterCache.OnResourceUpdated(func(newRes *clustercache.Resource, oldRes *clustercache.Resource, namespaceResources map[kube.ResourceKey]*clustercache.Resource) {
+		c.lock.RLock()
+		cacheSettings := c.cacheSettings
+		c.lock.RUnlock()
+
+		if cacheSettings.ignoreResourceUpdatesEnabled && oldRes != nil && newRes != nil && skipResourceUpdate(resInfo(oldRes), resInfo(newRes)) {
+			return
+		}
+
+		// TODO: when switch to change event triggering for sync tasking, notify
+		//       the Numaplane managing the resources that there are changes happen
```

Comment:
is the "event triggering" that this mentions essentially what we'll be doing through our watch which calls `Reconcile()`?

Reply:
yeah in the logged issue, specified we can do **Watch** first and use this later for improvements